src/confetti.js

```diff
@@ -281,18 +281,11 @@
     return colors.map(hexToRgb);
   }
 
-  function hexToRgb(str) {
-    var val = String(str).replace(/[^0-9a-f]/gi, '');
-
-    if (val.length < 6) {
-        val = val[0]+val[0]+val[1]+val[1]+val[2]+val[2];
-    }
-
-    return {
-      r: toDecimal(val.substring(0,2)),
-      g: toDecimal(val.substring(2,4)),
-      b: toDecimal(val.substring(4,6))
-    };
+  function hexToRgb(color) {
+    const ctx = document.createElement("canvas").getContext("2d");
+    ctx.fillStyle = color; // Set the fill style to the input color
+    const [r, g, b] = ctx.fillStyle.match(/\w\w/g).map(hex => parseInt(hex, 16)); // Convert to RGB
+    return { r, g, b };
   }
```

Comment:
```suggestion
  var canvas = typeof OffscreenCanvas === 'function' ? new OffscreenCanvas(0, 0) : document.createElement('canvas');
  var ctx = canvas.getContext('2d');
  var colorCache = Object.create(null);

  function colorToRgb(color) {
    if (colorCache[color] != null) return colorCache[color];

    ctx.fillStyle = color;
    var match = ctx.fillStyle.match(/^#([0-9a-f]{2})([0-9a-f]{2})([0-9a-f]{2})$/i);

    if (match == null) throw new Error('Unsupported color value: ' + color);

    var rgb = match.slice(1).map(function(hex) { return parseInt(hex, 16); });
    return colorCache[color] = { r: rgb[0], g: rgb[1], b: rgb[2] };
  }
```

Perf:
1. Re-use the same canvas/ctx rather than initializing each time
2. Use OffscreenCanvas instead of canvas HTML element where available
3. Cache color strings that have already been calculated

Strictness:

Throw a descriptive error in case the provided `color` string doesn't convert to a result in `#xxxxxx` format (e.g. colors with an alpha channel or using other color spaces — see https://drafts.csswg.org/css-color/#color-serialization-html-compatible-serialization-is-requested)

ES5/compat:

Not sure if it's really still required, but the rest of the library (including eslint configs) seems pretty conservative about it, so should probably avoid modern JS features/APIs here for consistency if nothing else.